### ISSUE TITLE
About how does buffer size from Dataset.shuffle() influence Dataset.cache()

### ISSUE THREAD

**Issue by xjlupeter** [tensorflow/tensorflow#55361]
Hi,
I find when I set `train_ds = train_ds.cache(filename='...')`
`train_ds = train_ds.shuffle(buffer_size=int(training_size/10)).batch(batch_size).prefetch(buffer_size=AUTOTUNE)`
During the first epoch of model fitting, what happened looks strange.

`Epoch 1/10
2022-03-24 16:54:22.982546: I tensorflow/stream_executor/cuda/cuda_dnn.cc:366] Loaded cuDNN version 8101
245/274 [=========================>....] - ETA: 8s - loss: 1.5153 - accuracy: 0.5639`

`2022-03-24 16:55:42.858875: W tensorflow/core/kernels/data/cache_dataset_ops.cc:233] The calling iterator did not fully read the dataset being cached. In order to avoid unexpected truncation of the dataset, the partially cached contents of the dataset  will be discarded. This can happen if you have an input pipeline similar to `dataset.cache().take(k).repeat()`. You should use `dataset.take(k).cache().repeat()` instead.`

`274/274 [==============================] - 109s 352ms/step - loss: 1.4703 - accuracy: 0.5753 - val_loss: 2.4110 - val_accuracy: 0.3950`

`Epoch 2/10
274/274 [==============================] - 93s 336ms/step - loss: 0.9126 - accuracy: 0.7168 - val_loss: 0.7763 - val_accuracy: 0.7676`

`Epoch 3/10
274/274 [==============================] - 92s 334ms/step - loss: 0.7067 - accuracy: 0.7886 - val_loss: 0.6996 - val_accuracy: 0.7922`
`

It seems that the model fitting ends before the feeding of the last 1/10 batches (this proportion is same as the proportion used in buffer size, I set this number in buffer size to save memory). Then the model fitting is restarted and the first epoch model fitting is finish.

I would like to ask 1) why this happens? 2) Is all my dataset successfully been cached rather than only the dataset for the first 90% batches is cached?

**Comment by tilakrayal** [comment]
@xjlupeter ,
Can you please take a look at this SO [link](https://stackoverflow.com/questions/64372390/what-does-buffer-size-do-in-tensorflow-dataset-shuffling) and [link](https://www.tensorflow.org/api_docs/python/tf/data/Dataset#shuffle) which delivers the required information.It helps.Thanks!
